Flaky test in generate_cert

This is flaky, and seems to be related to the cert timestamp.

W0831 04:01:56.424] PASS: //pkg/pki:go_default_test
W0831 04:01:56.546] ____[473 / 789] Testing //pkg/pki/ca:go_default_test
W0831 04:01:56.932] FAIL: //pkg/pki/ca:go_default_test (see /home/bootstrap/.cache/bazel/_bazel_bootstrap/82c87f91d10397dd2101a6893e4c013b/execroot/__main__/bazel-out/local-fastbuild/testlogs/pkg/pki/ca/go_default_test/test.log)
W0831 04:01:56.932] ____From Testing //pkg/pki/ca:go_default_test:
I0831 04:01:57.032] ==================== Test output for //pkg/pki/ca:go_default_test:
I0831 04:01:57.033] E0831 04:01:56.865231       9 generate_cert.go:82] RSA key generation failed with error crypto/rsa: too few primes of given length to generate an RSA key.
I0831 04:01:57.033] --- FAIL: TestGenCert (0.04s)
I0831 04:01:57.033] 	generate_cert_test.go:82: now: 2017-08-31 04:01:57 +0000 UTC
I0831 04:01:57.033] 	generate_cert_test.go:109: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.034] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.034] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.034] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.034] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.034] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.035] 	generate_cert_test.go:277: failed to verify certificate: x509: certificate has expired or is not yet valid
I0831 04:01:57.035] FAIL
I0831 04:01:57.035] Use --strategy=TestRunner=standalone to disable sandboxing for the failing actions.

--- a/pkg/pki/ca/generate_cert_test.go
+++ b/pkg/pki/ca/generate_cert_test.go
@@ -76,7 +76,7 @@ func TestGenCSRWithInvalidOption(t *testing.T) {
 }
 
 func TestGenCert(t *testing.T) {
-	// set "notBefore" to be 5 minutes aga, this ensures the issued certifiate to
+	// set "notBefore" to be 5 minutes ago, this ensures the issued certifiate to
 	// be valid as of now.
 	caCertNotBefore := now.Add(-5 * time.Minute)
 	caCertNotAfter := now.Add(24 * time.Hour)

--- a/pkg/pki/ca/generate_cert_test.go
+++ b/pkg/pki/ca/generate_cert_test.go
@@ -76,7 +76,9 @@ func TestGenCSRWithInvalidOption(t *testing.T) {
 }
 
 func TestGenCert(t *testing.T) {
-	caCertNotBefore := now
+	// set "notBefore" to be 5 minutes aga, this ensures the issued certifiate to
+	// be valid as of now.
+	caCertNotBefore := now.Add(-5 * time.Minute)
 	caCertNotAfter := now.Add(24 * time.Hour)
 
 	t.Logf("now: %+v\n", now)
@@ -119,6 +121,8 @@ func TestGenCert(t *testing.T) {
 		t.Error(err)
 	}
 
+	notAfter := now.Add(time.Hour)
+	notBefore := now.Add(-5 * time.Minute)
 	cases := []struct {
 		certOptions  CertOptions
 		verifyFields *tu.VerifyFields
@@ -128,8 +132,8 @@ func TestGenCert(t *testing.T) {
 			// server cert with DNS as SAN
 			certOptions: CertOptions{
 				Host:         "test_server.com",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 24),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -143,8 +147,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 24),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},
@@ -152,8 +156,8 @@ func TestGenCert(t *testing.T) {
 			// client cert with DNS as SAN
 			certOptions: CertOptions{
 				Host:         "test_client.com",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 36),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -167,8 +171,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 36),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},
@@ -176,8 +180,8 @@ func TestGenCert(t *testing.T) {
 			// server cert with IP as SAN
 			certOptions: CertOptions{
 				Host:         "1.2.3.4",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 24),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -191,8 +195,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 24),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},
@@ -200,8 +204,8 @@ func TestGenCert(t *testing.T) {
 			// client cert with service account as SAN
 			certOptions: CertOptions{
 				Host:         "spiffe://domain/ns/bar/sa/foo",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 100),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -215,8 +219,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 100),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},
@@ -224,8 +228,8 @@ func TestGenCert(t *testing.T) {
 			// server cert with service account as SAN
 			certOptions: CertOptions{
 				Host:         "spiffe://domain/ns/bar/sa/foo",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 50),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -239,8 +243,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 50),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},
@@ -248,8 +252,8 @@ func TestGenCert(t *testing.T) {
 			// a cert that can only be used as client-side cert
 			certOptions: CertOptions{
 				Host:         "spiffe://domain/ns/bar/sa/foo",
-				NotBefore:    now,
-				NotAfter:     now.Add(time.Hour * 50),
+				NotBefore:    notBefore,
+				NotAfter:     notAfter,
 				SignerCert:   caCert,
 				SignerPriv:   caPriv,
 				Org:          "",
@@ -263,8 +267,8 @@ func TestGenCert(t *testing.T) {
 				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
 				IsCA:        false,
 				KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-				NotAfter:    now.Add(time.Hour * 50),
-				NotBefore:   now,
+				NotAfter:    notAfter,
+				NotBefore:   notBefore,
 				Org:         "MyOrg",
 			},
 		},

